src/s3/commands/uploadFile.ts

```diff
@@ -26,51 +33,117 @@ export interface FileSizeBytes {
 }
 
 /**
- * Uploads a file to the bucket or folder represented by the given node.
+ * Wizard to upload a file.
+ * Either a node or the document has to be given, will follow different path depending on what is given
  *
- * Prompts the user for the file location.
- * Shows the output channel with "upload started" message.
- * Uploads the file (showing a progress bar).
- * Shows the output channel with "upload completed" message.
- * Refreshes the node.
+ * @param s3Client account to upload the file to
+ * @param nodeOrDocument node to upload to or file currently open
  *
- * Node that the node is reset to displaying its first page of results.
- * The file that is uploaded won't necessary fall on the first page.
- * The user may need to load more pages to see the uploaded file reflected in the tree.
  */
 export async function uploadFileCommand(
-    node: S3BucketNode | S3FolderNode,
+    s3Client: S3Client,
+    nodeOrDocument: S3BucketNode | S3FolderNode | vscode.Uri | undefined,
     fileSizeBytes: FileSizeBytes = statFile,
+    getBucket = promptUserForBucket,
+    getFile = getFileToUpload,
     window = Window.vscode(),
-    commands = Commands.vscode(),
-    outputChannel = ext.outputChannel
+    outputChannel = ext.outputChannel,
+    commands = Commands.vscode()
 ): Promise<void> {
-    getLogger().debug('UploadFile called for %O', node)
+    let key: string
+    let bucket: S3.Bucket
+    let file: vscode.Uri | undefined
+    let node: S3BucketNode | S3FolderNode | undefined
+    let document: vscode.Uri | undefined
 
-    const fileLocation = await promptForFileLocation(window)
-    if (!fileLocation) {
-        getLogger().info('UploadFile cancelled')
-        telemetry.recordS3UploadObject({ result: 'Cancelled' })
-        return
+    if (nodeOrDocument) {
+        if ((nodeOrDocument as any).getChildren) {
```

Comment:
What if the object wasn't created using new? Can that be a possibility?

Reply:
> What if the object wasn't created using new? Can that be a possibility?

For our production code no. They are all classes so they will be created using `new`. Of course, this is JavaScript so in the end it doesn't make much of a difference since callers can do whatever they want, that's why I marked the comment as a nit-pick. 

Using `instanceof` is just nice since it doesn't make assumptions about the specific properties of an object. The biggest flaw when using it is you can't easily 'fake' the objects you're checking when creating tests, but that's not applicable here. Pros and cons with both ways basically.